restapi/stories.go

```diff
@@ -57,12 +73,74 @@ func (api api) CreateStories(w http.ResponseWriter, r *http.Request) render.Rend
 
 	for i := 0; i < len(stories); i++ {
 		city := stories[i].CurrentCity
-		coordinates, err := api.locationFinder.GetCityCenter(city)
+		stories[i].CurrentCity = strings.Title(strings.ToLower(city))
+
+		// get number of stories with current city in db
+		var prevStories []models.Story
+		err := api.database.Where("current_city=?", city).Find(&prevStories).Error
+		if err != nil {
+			api.logger.Infof("A")
+			// default to randomizing if there is an error
+			stories[i].Latitude, stories[i].Longitude, err = api.randomizePinLocation(city)
+			if err != nil {
+				return rest.ErrInvalidRequest(api.logger, fmt.Sprintf("Story %d has an invalid current city", i), err)
+			}
+			continue
+		}
+		numStoriesInCity := 0
+		if prevStories != nil {
+			numStoriesInCity = len(prevStories)
+		}
+		// query zipcode with numStoriesInCity + 1 results and take the last one
+		zipCodeToken := os.Getenv("ZIP_CODE_TOKEN")
+		resp, err := http.Get(fmt.Sprintf("https://app.zipcodebase.com/api/v1/code/city?apikey=%s&city=%s&country=ca&limit=%d", zipCodeToken, city, numStoriesInCity+1))
+		if err != nil || resp.StatusCode != 200 {
+			api.logger.Infof("B")
+			// default to randomizing if there is an error
+			stories[i].Latitude, stories[i].Longitude, err = api.randomizePinLocation(city)
+			if err != nil {
+				return rest.ErrInvalidRequest(api.logger, fmt.Sprintf("Story %d has an invalid current city", i), err)
+			}
+			continue
+		}
+		defer resp.Body.Close()
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			api.logger.Infof("C")
+			// default to randomizing if there is an error
+			stories[i].Latitude, stories[i].Longitude, err = api.randomizePinLocation(city)
+			if err != nil {
+				return rest.ErrInvalidRequest(api.logger, fmt.Sprintf("Story %d has an invalid current city", i), err)
+			}
+			continue
+		}
+		var zipcodeResp ZipcodeResponse
+		err = json.Unmarshal(body, &zipcodeResp)
+		if err != nil || zipcodeResp.Results == nil || len(zipcodeResp.Results) == 0 {
+			api.logger.Infof("D")
+			// default to randomizing if there is an error
+			stories[i].Latitude, stories[i].Longitude, err = api.randomizePinLocation(city)
+			if err != nil {
+				return rest.ErrInvalidRequest(api.logger, fmt.Sprintf("Story %d has an invalid current city", i), err)
+			}
+			continue
+		}
+
+		postalCode := zipcodeResp.Results[len(zipcodeResp.Results)-1]
```

Comment:
Why the last element of the array?

Reply:
Esentially the number of postal codes we request from the API is how many stories we currently have + 1 (for the new story). So we get the last postal code because it is the newest (the API always returns postal codes in the same order so we can't request 1 each time otherwise we would get the same one).

Also we can't load all postal codes each time otherwise we'd hit the api limit in 2-3 api requests.